File: android/app/src/main/kotlin/io/github/vanzeph/minutrove/BackupFiles.kt
package io.github.vanzeph.minutrove

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import androidx.core.content.FileProvider
import io.flutter.plugin.common.MethodChannel
import java.io.File

/**
 * Owns the OS file interactions behind the Dart [BackupFilePicker] and
 * [BackupFileSharer] interfaces: one `.minutrove` document pick for restore
 * and one share-sheet hand-off for export.
 *
 * Both operations are user-driven UI, so an answer arrives only after the
 * user closes the presented sheet. Cancelling the picker answers null;
 * nothing is read and nothing changes. Sharing stages the immutable export
 * in the app's cache only for the hand-off and removes it afterwards.
 *
 * FlutterActivity extends the framework Activity, so the picker uses the
 * classic [Activity.startActivityForResult] round trip; the owning activity
 * forwards [onActivityResult] here.
 */
class BackupFiles(private val activity: Activity) {
    companion object {
        const val channelName = "io.github.vanzeph.minutrove/files"
        const val fileExtension = "minutrove"
        const val pickRequestCode = 704617
    }

    private var pendingResult: MethodChannel.Result? = null

    fun handle(call: MethodChannel.MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "pickBackup" -> {
                if (pendingResult != null) {
                    result.error("pick_already_active", "A file picker is already open", null)
                    return
                }
                pendingResult = result
                val intent = Intent(Intent.ACTION_OPEN_DOCUMENT).apply {
                    addCategory(Intent.CATEGORY_OPENABLE)
                    type = "*/*"
                    // Every storage provider filters by declared extras; the
                    // generic data type keeps a renamed file selectable.
                    putExtra(Intent.EXTRA_MIME_TYPES, arrayOf("*/*"))
                }
                activity.startActivityForResult(intent, pickRequestCode)
            }
            "shareBackup" -> {
                val fileName = call.argument<String>("fileName")
                val bytes = call.argument<ByteArray>("bytes")
                if (fileName.isNullOrEmpty() || fileName.length > 128 ||
                    !fileName.endsWith(".$fileExtension") || fileName.contains("/") ||
                    bytes == null
                ) {
                    result.error("invalid_share_request", "A .minutrove file name and bytes are required", null)
                    return
                }
                share(fileName, bytes, result)
            }
            else -> result.notImplemented()
        }
    }

    /** Returns true when the result belongs to the backup picker. */
    fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != pickRequestCode) return false
        val result = pendingResult
        pendingResult = null
        if (result == null) return true
        val uri = data?.data
        if (resultCode != Activity.RESULT_OK || uri == null) {
            // Cancellation is an expected answer, not an error.
            result.success(null)
            return true
        }
        try {
            val bytes = activity.contentResolver.openInputStream(uri)?.use { it.readBytes() }
            if (bytes == null) {
                result.error("pick_read_failed", "Could not read the selected file", null)
            } else {
                result.success(bytes)
            }
        } catch (_: Exception) {
            result.error("pick_read_failed", "Could not read the selected file", null)
        }
        return true
    }

    private fun share(fileName: String, bytes: ByteArray, result: MethodChannel.Result) {
        try {
            val directory = File(activity.cacheDir, "exports").apply { mkdirs() }
            // A stale staged file from an interrupted share must not be handed out.
            directory.listFiles()?.forEach { it.delete() }
            val staged = File(directory, fileName)
            staged.writeBytes(bytes)
            val uri = FileProvider.getUriForFile(
                activity, "${activity.packageName}.fileprovider", staged)
            val intent = Intent(Intent.ACTION_SEND).apply {
                type = "application/octet-stream"
                putExtra(Intent.EXTRA_STREAM, uri)
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }
            activity.startActivity(Intent.createChooser(intent, fileName))
            result.success(true)
        } catch (_: ActivityNotFoundException) {
            result.error("share_unavailable", "No app can share this file", null)
        } catch (_: Exception) {
            result.error("share_failed", "Could not stage the backup for sharing", null)
        }
    }
}

File: ios/Runner/BackupFiles.swift
import Flutter
import Foundation
import UniformTypeIdentifiers
import UIKit

/// Owns the OS file interactions behind the Dart `BackupFilePicker` and
/// `BackupFileSharer` interfaces: one `.minutrove` document pick for restore
/// and one share-sheet hand-off for export.
///
/// Both operations are user-driven UI, so an answer arrives only after the
/// user closes the presented sheet. Cancelling the picker answers `nil`;
/// nothing is read and nothing changes. Reading uses a coordinated file
/// access; bytes are answered once and never cached. Sharing writes the
/// immutable export to an OS-owned temporary location only for the hand-off.
final class BackupFiles {
  static let channelName = "io.github.vanzeph.minutrove/files"
  static let fileExtension = "minutrove"

  /// One in-flight operation at a time: a second request while a picker is
  /// presented is rejected instead of dropping the first answer.
  private var pendingPickerResult: FlutterResult?

  // MARK: Channel operations

  func pickBackup(result: @escaping FlutterResult) {
    DispatchQueue.main.async {
      guard self.pendingPickerResult == nil else {
        result(FlutterError(code: "pick_already_active",
                            message: "A file picker is already open", details: nil))
        return
      }
      guard let presenter = UIApplication.shared.keyWindowRootController else {
        result(FlutterError(code: "picker_unavailable",
                            message: "No presenter is available", details: nil))
        return
      }
      self.pendingPickerResult = result
      // Prefer the declared extension; fall back to generic data so a renamed
      // or not-yet-registered file can still be chosen and validated in Dart.
      let types: [UTType] = [UTType(filenameExtension: BackupFiles.fileExtension,
                                    conformingTo: .data) ?? .data, .data]
      let picker = UIDocumentPickerViewController(forOpeningContentTypes: types, asCopy: false)
      picker.delegate = self
      picker.modalPresentationStyle = .pageSheet
      presenter.present(picker, animated: true)
    }
  }

  func shareBackup(fileName: String, bytes: FlutterStandardTypedData, result: @escaping FlutterResult) {
    DispatchQueue.main.async {
      guard !fileName.isEmpty, fileName.count <= 128,
            fileName.hasSuffix(".\(BackupFiles.fileExtension)"),
            !fileName.contains("/") else {
        result(FlutterError(code: "invalid_share_request",
                            message: "A .minutrove file name is required", details: nil))
        return
      }
      guard let presenter = UIApplication.shared.keyWindowRootController else {
        result(FlutterError(code: "share_unavailable",
                            message: "No presenter is available", details: nil))
        return
      }
      do {
        let url = FileManager.default.temporaryDirectory.appendingPathComponent(fileName)
        try? FileManager.default.removeItem(at: url)
        try bytes.data.write(to: url, options: .atomic)
        let controller = UIActivityViewController(activityItems: [url], applicationActivities: nil)
        controller.completionWithItemsHandler = { (_, _, _, _) in
          try? FileManager.default.removeItem(at: url)
        }
        if let popover = controller.popoverPresentationController {
          popover.sourceView = presenter.view
          popover.sourceRect = presenter.view.bounds.insetBy(dx: presenter.view.bounds.width / 4,
                                                              dy: presenter.view.bounds.height / 4)
        }
        presenter.present(controller, animated: true) {
          result(true)
        }
      } catch {
        result(FlutterError(code: "share_write_failed",
                            message: "Could not stage the backup for sharing", details: nil))
      }
    }
  }
}

extension BackupFiles: UIDocumentPickerDelegate {
  func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
    guard let result = pendingPickerResult else { return }
    pendingPickerResult = nil
    guard let url = urls.first else {
      result(nil)
      return
    }
    let coordinated = NSFileCoordinator()
    var bytes: Data?
    var coordinateError: NSError?
    coordinated.coordinate(readingItemAt: url, options: .forUploading,
                           error: &coordinateError) { readingURL in
      bytes = try? Data(contentsOf: readingURL)
    }
    if let bytes {
      result(FlutterStandardTypedData(bytes: bytes))
    } else {
      result(FlutterError(code: "pick_read_failed",
                          message: "Could not read the selected file", details: nil))
    }
  }

  func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
    guard let result = pendingPickerResult else { return }
    pendingPickerResult = nil
    result(nil)
  }
}

extension UIApplication {
  /// The controller that currently presents the app's visible interface; the
  /// Flutter view controller when no custom sheet is above it.
  var keyWindowRootController: UIViewController? {
    connectedScenes.compactMap { scene -> UIViewController? in
      guard let windowScene = scene as? UIWindowScene,
            scene.activationState == .foregroundActive || scene.activationState == .foregroundInactive
      else { return nil }
      return windowScene.keyWindow?.rootViewController
    }.first ?? connectedScenes.compactMap { ($0 as? UIWindowScene)?.keyWindow?.rootViewController }.first
  }
}
